pkg/util/list.go

```diff
@@ -0,0 +1,108 @@
+package util
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strconv"
+)
+
+// ListIterator lists the objects from the given url by loading one page at a time
+type ListIterator struct {
+	// DoRequest function executes the HTTP request, it is responsible for authentication
+	DoRequest DoRequestFunc
+	// URL is the address of the resource to list
+	URL string
+
+	next string
+	done bool
+}
+
+type listResponse struct {
+	Token    string      `json:"token"`
+	NumItems int64       `json:"num_items"`
+	Items    interface{} `json:"items"`
+}
+
+// Next loads the next page of items
+// items should be a pointer to a slice, that will be populated with the items from the current page,
+// if nil, only the total number is returned
+// maxItems is the maximum numbe of items to load with the next page, -1 - use server default, 0 - just get the count
+// more is true if there are more items
+// count is the total number of items, -1 if not available
+func (li *ListIterator) Next(ctx context.Context, items interface{}, maxItems int) (more bool, count int64, err error) {
```

Comment:
using raw interfaces is against the current design idea for SM. Also, Go best practices advice to accept interfaces and we have designed the Object interface for this purpose. Can we try to accept ObjectList or []Object here and remove all the reflection ?